### PR TITLE
Batched up spec updates

### PR DESCRIPTION
**Pull request by leafpetersen**
This batches up a number of clarifications and changes that have accumulated.  

cc @lrhn @munificent @eernstg @scheglov @stefantsov @liamappelbe @crelier @nshahan @fishythefish @sigmundch @johnniwinther

### CODE REVIEW

File: accepted/future-releases/nnbd/feature-specification.md [hunk 1]
```diff
@@ -412,6 +427,54 @@ Tearing off a method from a receiver of static type `Never` produces a value of
 type `Never`.  Applying an expression of type `Never` in the function position
 of a function call produces a result of type `Never`.
 
+The static type of a `throw e` expression is `Never`.
+
+In legacy mode, an override of operator== with no explicit parameter type
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: @johnniwinther @scheglov

File: accepted/future-releases/nnbd/feature-specification.md [hunk 2]
```diff
@@ -412,6 +427,54 @@ Tearing off a method from a receiver of static type `Never` produces a value of
...
+type.  Otherwise, the parameter type of the overriding method is dynamic.
+
+Top level variable and local function inference is performed
+as
+[specified separately](https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md).
+Method body inference is not yet specified.
+
+### Instantiate to bounds
+
+The computation of instantiation to bounds is changed to substitute `Never` for
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: @eernstg Is this what we want here?

File: accepted/future-releases/nnbd/feature-specification.md [hunk 3]
```diff
@@ -412,6 +427,54 @@ Tearing off a method from a receiver of static type `Never` produces a value of
...
+type variables appearing in contravariant positions instead of `Null`.
+
+### Least and greatest closure
+
+The definitions of least and greatest closure are changed in null safe libraries
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: cc @scheglov @stefantsov note this change

File: accepted/future-releases/nnbd/feature-specification.md [hunk 4]
```diff
@@ -990,6 +1053,10 @@ as:
    - And the reverse
  - `NNBD_TOP_MERGE(Object?, dynamic)  = Object?`
    - And the reverse
+ - `NNBD_TOP_MERGE(Object*, void)  = void`
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: @stefantsov Note this fix.

File: accepted/future-releases/nnbd/feature-specification.md [hunk 5]
```diff
@@ -1206,3 +1283,13 @@ understand these changes: for example, by providing a debugging option in which
 instance checks or casts which would result in a different outcome if run in
 strong checking mode vs weak checking mode are flagged for the developer by
 logging a warning or breaking to the debugger.
+
+### Automatic debug assertion insertion
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: @johnniwinther @stefantsov This change is lower priority, but something that I'd like to do if we have time before the release.

File: resources/type-system/subtyping.md [hunk 6]
```diff
@@ -92,11 +92,7 @@ capture-avoiding substitution of types `T0, ..., Tk` for the type variables
 ## Type equality
 
 We say that a type `T0` is equal to another type `T1` (written `T0 === T1`) if
-the two types are structurally equal up to renaming of bound type variables,
-and equating all top types.
-
-TODO: make these rules explicit.
-
+`T0 <: T1` and `T1 <: T0`.
```
Comment (leafpetersen) [inline resources/type-system/subtyping.md]: cc @stefantsov @johnniwinther @fishythefish @nshahan @liamappelbe @crelier @scheglov This change makes the subtyping rules for generic methods more forgiving.  It is specifically intended to allow a generic method with a legacy bound to override one with a nullable or non-nullable bound.  It also has the effect of allowing overrides that use different variants of top types for bounds. It also accounts for ad hoc normalization (e.g. `T Function<T extends int?>()` is specified to be a subtype of `T Function<T extends int??>()` by this change, and hence this validates the normalization of `int??` to `int?`).

File: accepted/future-releases/nnbd/feature-specification.md [hunk 7]
```diff
@@ -6,6 +6,23 @@ Status: Draft
 
 ## CHANGELOG
 
+2020.01.29
+  - **CHANGE** Relax the exhaustiveness check on switches.
+  - Specify the type of throw expressions.
+  - Specify the override inference exception for operator==.
+  - **CHANGE** Specify that instantiate to bounds uses `Never` instead of `Null.
```
Comment (kevmoo) [inline accepted/future-releases/nnbd/feature-specification.md]: missed trailing "tick" after `Null` on this line!

Comment (leafpetersen) [reply accepted/future-releases/nnbd/feature-specification.md]: Fixed, thanks!

File: resources/type-system/normalization.md [hunk 8]
```diff
@@ -80,7 +80,7 @@ equations apply gives us something like the following:
 - **NORM**(`X extends T`) =
   - let `S` be **NORM**(`T`)
   - if `S` is `Never` then `Never`
-  - else `X extends S`
+  - else `X extends T`
```
Comment (fishythefish) [inline resources/type-system/normalization.md]: Why this change? Presumably, `X extends int??` (for example) would be normalized to `X extends int?`, but IIUC this change is undoing that.

File: resources/type-system/subtyping.md (hunk 6 above)
Comment (fishythefish) [reply resources/type-system/subtyping.md]: Is this change gated behind anything, or can we just update the (legacy) subtyping algorithm?

File: resources/type-system/normalization.md (hunk 8 above)
Comment (leafpetersen) [reply resources/type-system/normalization.md]: Note the difference between normalizing a *use* of a type variable and the *declaration* of a type variable.  This case covers type variable uses. It's not clear to me what it means to normalize the bound on a use.  It's fine to do so, if a representation is chosen in which each use has a copy of the bound, but the bound does not contribute to syntactic equality.   The bound does contribute to normalization, since `X extends Never` normalizes to `Never`.  Does that make sense?

File: resources/type-system/subtyping.md (hunk 6 above)
Comment (leafpetersen) [reply resources/type-system/subtyping.md]: Let's just update in place.

File: resources/type-system/normalization.md (hunk 8 above)
Comment (fishythefish) [reply resources/type-system/normalization.md]: Sort of. I'm reading the various expressions in this document as (abstract) type terms, and in my mental model, these aren't intrinsically equipped with a notion of being declared vs. used. (In a Dart program, type terms certainly exist in declaration-contexts and use-contexts, but I'm interpreting normalization as a context-free operation.)

Maybe I missed a definition somewhere, but I'm not really sure how to read `X extends T` as a standalone type term. (That is, I recognize it as a *declaration* in the context of a type parameter list, but I'm not sure how to interpret an `extends` clause without that context the way I can interpret `FutureOr<X>`, for example.) In this case, I took it as shorthand for the following: I have a type variable `X` and when I query my type environment's oracle/function/mechanism for looking up bounds, it tells me that **bound**(`X`) = `T`.

Comment (leafpetersen) [reply resources/type-system/normalization.md]: > In this case, I took it as shorthand for the following: I have a type variable `X` and when I query my type environment's oracle/function/mechanism for looking up bounds, it tells me that **bound**(`X`) = `T`.

Functionally, that is the correct interpretation.  The analyzer's representation doesn't really work that way.  Instead, they attach a bound to every type variable, and there's essentially a coherence property that you enforce.  But it's morally the same thing - it's just a question of whether you keep a map from type variables to bounds, or just enforce that all type variables bound at the same location have the same bound attached to them.  I've just gotten in the habit of using the `X extends T` notation instead of talking about "the bound associated with T in the environment" since that also matches up better with the implementations.

File: accepted/future-releases/nnbd/feature-specification.md [hunk 9]
```diff
@@ -752,10 +815,9 @@ implicitly induced by _D_ if a value has previously been assigned to `v`
 (which could be due to an initializing formal or a constructor initializer
 list, or due to an invocation of the setter).
 
-Let _D_ be a `late` and `final` local variable declaration named `v`
-without an initializing expression.  It is a run-time error, throwing an
-instance of `LateInitializationError`, to assign a value to `v` if a value
-has previously been assigned to `v`.
+Let _D_ be a `late` and `final` local variable declaration named `v`.  It is a
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: cc @johnniwinther

**Comment by leafpetersen** [comment]
Landing this for now, I can iterate based on feedback in followups.
